### PR TITLE
Switch from assertions to invariants/preconditions in exception handling

### PR DESCRIPTION
**Pull request by cristina-david**

### CODE REVIEW

File: src/goto-programs/remove_exceptions.cpp [hunk 1]
```diff
@@ -116,7 +118,9 @@ void remove_exceptionst::add_exceptional_returns(
     {
       const exprt &function_expr=
         to_code_function_call(instr_it->code).function();
-      assert(function_expr.id()==ID_symbol);
+      INVARIANT(
+        function_expr.id()==ID_symbol,
```
Comment (martin-cs) [inline src/goto-programs/remove_exceptions.cpp]: Likewise, is this a DATA_INVARIANT, i.e. a condition on the general well-structured-ness of the exprt?

File: src/analyses/uncaught_exceptions_analysis.cpp [hunk 2]
```diff
@@ -122,7 +122,9 @@ void uncaught_exceptions_domaint::transform(
   {
     const exprt &function_expr=
       to_code_function_call(instruction.code).function();
-    assert(function_expr.id()==ID_symbol);
+    INVARIANT(
+      function_expr.id()==ID_symbol,
```
Comment (martin-cs) [inline src/analyses/uncaught_exceptions_analysis.cpp]: Is this actually a DATA_INVARIANT?

File: src/goto-programs/remove_exceptions.cpp [hunk 3]
```diff
@@ -226,9 +232,13 @@ static goto_programt::targett get_exceptional_output(
     const irep_idt &statement=it->code.get_statement();
     if(statement==ID_output)
     {
-      assert(it->code.operands().size()>=2);
+      INVARIANT(
+        it->code.operands().size()>=2,
+        "Output expected to have at least 2 operands");
```
Comment (martin-cs) [inline src/goto-programs/remove_exceptions.cpp]: Is definitely a DATA_INVARIANT.

File: src/goto-programs/remove_exceptions.cpp [hunk 4]
```diff
@@ -226,9 +232,13 @@ static goto_programt::targett get_exceptional_output(
...
       const exprt &expr=it->code.op1();
-      assert(expr.id()==ID_symbol);
+      INVARIANT(
+        expr.id()==ID_symbol,
```
Comment (martin-cs) [inline src/goto-programs/remove_exceptions.cpp]: Probably likewise.

File: src/goto-programs/remove_exceptions.cpp (hunk 3 above)
Comment (peterschrammel) [inline src/goto-programs/remove_exceptions.cpp]: Nitpicking on uniformity: Is there a guideline on whether the message starts lower or upper case?

File: src/analyses/uncaught_exceptions_analysis.cpp [hunk 5]
```diff
@@ -122,7 +122,9 @@ void uncaught_exceptions_domaint::transform(
...
+      "Identifier expected to be a symbol");
```
Comment (peterschrammel) [inline src/analyses/uncaught_exceptions_analysis.cpp]: https://github.com/diffblue/cbmc/blob/test-gen-support/CODING_STANDARD#L157

**Comment by cristina-david** [comment]
This has been rebased and it's ready for a new round of reviews.

**Pull request merged**